### ISSUE TITLE
Issue with EMV Card Dip  

### ISSUE THREAD

**Issue by rishabhbsdei** [CardFlight/cardflight-android#10]
I have updated Cardflight  SDK for android to 3.2.3 . After that I am facing issues with EMV Card transaction while dipping  . I am getting "timeout" error in

public void readerCardResponse(Card card, CardFlightError cardFlightError) callback.

But according to your documentation,
public void emvTransactionResult(Charge charge, boolean b, CFEMVMessage cfemvMessage) method should be called for emvResult.

Apart from this it is also deducting amount from credit card.

According to your documentation it is expected to call 
emvTransactionResult() method while dipping but now I am receiving callback in readerCardResponse() which belongs to swipe feature . And I am facing this problem after updating my SDK to 3.2.3(Previous version was 3.1.1).

**Comment by austinodell** [comment]
Hi Rishabh,

Is the issue with timeout getting called in `readerCardResponse` happening before or after you call `beginEMVTransaction`?  And have you set `swipeHasTimeout(false)`?

And by the second issue, do you mean that the transactions are going through after the timeout error has been given?

The transaction result should be given in `emvTransactionResult` so long as you called `beginEMVTransaction`, and not `beginSwipe`.  Let me know which of these you are using.

Thanks,
Austin

**Comment by rishabhbsdei** [comment]
Hi Austin
I am calling beginEMVTransaction(intPrice, chargeDetailsHashMap) method after I get callback in emvCardDipped().
I am not using swipeHasTimeout().
The transactions are going before the timeout error appears in readerCardResponse() method. I am checking logs from Cardflight SDK as It create a Success response log and also amount is deducted at that time.
I am not calling beginSwipe in emvCardDipped method . I am calling beginSwipe method only in readerCardResponse and readerCardFailed to make transaction for second time . 
Please suggest me on my above response . And also ,Now I am unable to connect with A200 with the updated SDK.
Thanks
Rishabh

**Comment by rishabhbsdei** [comment]
Hello Austin
Please check my Logs below

06-01 07:52:47.219 13689-13689/com.parkjockey.operator E/Debug: readerIsConnecting 
06-01 07:53:04.216 13689-13689/com.parkjockey.operator E/Debug: readerIsAttached
06-01 07:53:04.226 13689-13689/com.parkjockey.operator E/Debug: readerIsConnected Eclipse A200
06-01 07:53:06.558 13689-13689/com.parkjockey.operator E/Debug: readerCardResponse - Card null

I am calling      Reader.getDefault(mContext).beginSwipe(); method only if Card Reader Type is A100 . So according to your documentation readerCardResponse only be called after that . But as in my above log if Device is A200 ,readerCardResponse is calling automatically after device connected and giving Card as null.
Please check this issue and also explain if I need to do something to dip card and make transactions.
Thanks 
Rishabh

**Comment by austinodell** [comment]
Hi Rishabh,

There is a known issue in this version of the SDK where `readerCardResponse(null, TIMEOUT)` is being called immediately after connecting to an A200. You can safely ignore this for now and it will be fixed in the next release.  Are you able to make transactions with the A200 if you ignore the timeout message?

Thanks,
Austin

**Comment by rishabhbsdei** [comment]
Hi Austin

Thanks for your reply.
I am not able to do transactions with the A200 if I ignore **readerCardResponse(null,TIMEOUT)** too. I also checked this in your latest sample code . It gave me success connection and success Apitoken for emv .Please check below logs from your sample app.

06-02 09:57:20.692 6157-6157/com.example.cardflight V/CardFlight: [cp.requestSuccessful] Account: is_emv_ready = true
06-02 09:57:20.693 6157-6157/com.example.cardflight I/ReaderDemoFragment: API Key and Account Token set!

But If I dip my credit card in A200 , its not calling any callback.
One more thing ,as A200 also support swipe feature so should I need to call beginSwipe() method before swiping my card ?
Please suggest me as soon as possible .This is very urgent for us.

Thanks
Rishabh

**Comment by austinodell** [comment]
Hi Rishabh,

Was the issue fixed after charging your A200?

You should be calling
```
Reader.getDefault(mContext)
    .beginEMVTransaction(int amount, HashMap chargeDetails);
```
before you are able to receive any EMV-related callbacks.

You can use `beginSwipe()` with the A200, but that will not allow EMV transactions and will only use the mag stripe on the reader.

Thanks,
Austin

**Comment by rishabhbsdei** [comment]
Hi Austin
The A200 connected with my app after charging.But I am unable  to receive any callback of emvCardDipped();
I think Reader.getDefault(mContext)
    .beginEMVTransaction(int amount, HashMap chargeDetails); should  be called after I receive callback in emvCardDipped():

Thanks
Rishabh

**Comment by rishabhbsdei** [comment]
I checked this with your Demo app as well ,according to your sample app an Alert(Snackbar) should be displayed if card dipped in machine . But its not happening at my end . So please suggest how to recognize  whether machine (A200) is working properly for dipping functionality or not .

Thanks
Rishabh

**Comment by austinodell** [comment]
Hi Rishabh,

The A200 does not start listening for card activity until a transaction has been started. It does this to save battery. So you won't receive an `emvCardDipped` callback until you've begun the transaction.  Once you receive that, you should update your UI to show the user that the process is working.  You don't need to make any other calls until you receive the `emvCardResponse(HashMap details)` callback. Then you should display a message to the user asking them to verify at a minimum the last four digits of the card and the amount (details also contains the first six). You can access these using the `Constants` class (e.g. `Constants.LAST_FOUR`). With the user's response, you should then call
```
Reader.getDefault(this)
    .emvProcessTransaction(response)
```
where `response` is a boolean representing the user's confirmation choice.  Once you have called that, you will receive the transaction result through the `emvTransactionResult` callback.

Keep in mind that throughout this whole process you should be listening for the `emvMessage` callback and displaying these to the user, as well as the other EMV-related callbacks which each require various actions to be taken.

Thanks,
Austin

**Comment by rishabhbsdei** [comment]
Hi Austin

As per your response
**The A200 does not start listening for card activity until a transaction has been started.** 
How can I start transaction ? I am receiving callback in     **public void readerIsConnected(boolean b, CardFlightError cardFlightError)**  after reader is connected to my application . What should I need to do so that it start listening emv related callbacks ?
Also  I am not getting message in **emvMessage** for emvCardDipped . Please suggest me for this. 
As per your previous SDK we have to just connect the Reader (A200) with our app and after that it automatically calls emvCardDipped() method if card was dipped.

Thanks 
Rishabh

**Comment by rishabhbsdei** [comment]
Austin , I am beginning the transaction by dipping the card into A200 reader and I was expecting  to call **emvCardDipped()**  automatically by reader . But this is not happening at my end not even in your sample app. As in your sample app you have provided "Attempt Swipe"Button to test the swipe feature , have you provide any button to check Card Dip feature in your sample app?

**Comment by austinodell** [comment]
Hi Rishabh,

To begin the transaction, you **must** call
```
Reader.getDefault(this)
    .beginEmvTransaction(amount, chargeDetails);
```
Before calling that, you will not receive any sort of EMV-related messaging.

The sample app does not implement EMV because it is not easy to fake like it is with keyed/swiped transactions.

Calling `beginSwipe` will not listen for EMV-type transactions either, so it must be `beginEmvTransaction`.

Thanks,
Austin

**Comment by rishabhbsdei** [comment]
Hi Austin
Thanks for your support.

As per your above response I called **beginEmvTransaction** method on my application screen only when reader is connected and Activity is start visible to user(in onResume() method of activity). After doing this the transaction is being proceeded and make a successful payment in positive testing . But if I tried to dip my card multiple times on the same screen without doing any successful transaction ,**emvCardDipped()** is calling multiple times in the third attempt .  It is happening in case if I cancel the transaction by removing card in between the transaction and again trying to dip the card without going back to my previous screen.
Please check this issue.

**Comment by austinodell** [comment]
Hi Rishabh,

I'm not sure if I fully understand what you mean here.  Are you trying to dip, remove, dip, remove, and dip again and expect the transaction to work?  If the card gets removed after the transaction has begun, it cancels the transaction and you will need to start it again to make it work.

Thanks,
Austin

**Comment by rishabhbsdei** [comment]
Hi Austin
I was trying to dip ,remove my card again trying to dip again remove my card . My transaction has been canceled ,this is correct thing . But when tried this for the third time (Dip card) then emvCardDipped() method was calling multiple time and it was not canceling that transaction even if I remove the card from machine.
Let me know if need more details .
And also please let me know about the bug fixes in your latest SDK(3.2.4) .Should we need to update this or we can work with previous one? Also is there any known issue in your latest SDK also?

**Comment by austinodell** [comment]
Hi Rishabh,

Are you calling `beginEmvTransaction` after the transaction is cancelled, or are you trying to dip using that same transaction?

SDK v3.2.4 contains a lot of small bug fixes, as well as fixes for the amount not being passed to specific readers under certain conditions, and a fix for Card Removed being shown more than it should.  It was a minor release, but it is always worth updating to the latest SDK to minimize bugs.  As of now, there are no known issues with v3.2.4.

Thanks,
Austin
